rig/routing_table/utils.py

```diff
@@ -0,0 +1,238 @@
+import warnings
+from rig.routing_table import RoutingTableEntry
+
+
+def table_is_subset_of(entries_a, entries_b):
+    """Check that every key matched by every entry in one table results in the
+    same route when checked against the other table.
+
+    For example::
+
+        >>> from rig.routing_table import Routes
+        >>> table = [
+        ...     RoutingTableEntry({Routes.north, Routes.north_east}, 0x0, 0xf),
+        ...     RoutingTableEntry({Routes.east}, 0x1, 0xf),
+        ...     RoutingTableEntry({Routes.south_west}, 0x5, 0xf),
+        ...     RoutingTableEntry({Routes.north, Routes.north_east}, 0x8, 0xf),
+        ...     RoutingTableEntry({Routes.east}, 0x9, 0xf),
+        ...     RoutingTableEntry({Routes.south_west}, 0xe, 0xf),
+        ...     RoutingTableEntry({Routes.north, Routes.north_east}, 0xc, 0xf),
+        ...     RoutingTableEntry({Routes.south, Routes.south_west}, 0x0, 0xb),
+        ... ]
+
+    Is a functional subset of a minimised version of itself::
+
+        >>> from rig.routing_table import minimise
+        >>> other_table = minimise(table, target_length=None)
+        >>> other_table == table
+        False
+        >>> table_is_subset_of(table, other_table)
+        True
+
+    But not vice-versa::
+
+        >>> table_is_subset_of(other_table, table)
+        False
+
+    Parameters
+    ----------
+    entries_a : [:py:class:`~rig.routing_table.RoutingTableEntry`, ...]
+    entries_b : [:py:class:`~rig.routing_table.RoutingTableEntry`, ...]
+        Ordered of lists of routing table entries to compare.
+
+    Returns
+    -------
+    bool
+        True if every key matched in `entries_a` would result in an equivalent
+        route for the packet when matched in `entries_b`.
+    """
+    class NoMatchError(Exception):
+        pass
+
+    def get_first_match(key):
+        """Get the route from the first entry which matches the key."""
+        for entry in entries_b:
+            if (entry.mask & key) == entry.key:
+                return entry.route
+        raise NoMatchError  # No match was found at all
+
+    # Check that all the routes from the expanded first table are equivalent in
+    # the second table. NB: the first table is expanded in all positions save
+    # for where there are common Xs in the second table.
+    common_xs = get_common_xs(entries_b)
+    try:
+        return all(get_first_match(entry.key) == entry.route for entry in
```

Comment:
As much as I like this style; this algorithm would certainly be clearer as a for-loop over the keys with the above function simply being inlined as a nested loop.